github.go

```diff
@@ -201,48 +232,96 @@ func unauthenticatedGet(pages int, page int, client *http.Client) (GithubEvents,
 			Timeout: time.Second * 30,
 		}
 	}
-	r, err := httpClient.Get("https://api.github.com/events?page=" + strconv.Itoa(page))
+	r, err := httpClient.Get(url)
 	if err != nil {
 		log.Fatalf("Error in requesting data from API: %s", err.Error())
 	}
-	defer r.Body.Close()
 
-	RateLimitSpecs.Limit = parseHeader(r.Header, "x-ratelimit-limit")
-	RateLimitSpecs.Remaining = parseHeader(r.Header, "x-ratelimit-remaining")
-	RateLimitSpecs.ResetTimestamp = parseLongHeader(r.Header, "x-ratelimit-reset")
-	RateLimitSpecs.PollInterval = parseHeader(r.Header, "x-poll-interval") * pages
+	defer r.Body.Close()
 
 	if r.StatusCode == http.StatusNotModified { // (304) No new content
 		return nil, &APIError{"no new content", 304}
 	} else if r.StatusCode == http.StatusForbidden { // (403) Rate limit reached
 		return nil, &APIError{"rate limited", 403}
 	}
 
-	body, _ := ioutil.ReadAll(r.Body)
+	body, err := ioutil.ReadAll(r.Body)
+
+	if err != nil {
+		log.Fatal("Unable to read response body")
+		return nil, err
+	}
+
+	return body, nil
+}
+
+func getGithubEvents(page int, pages int, token string) (GithubEvents, error) {
+	var responseBody []byte
+	var err error
+
+	if token == "" {
+		responseBody, err =
+			unauthenticatedGet("https://api.github.com/events?page="+strconv.Itoa(page), nil)
+	} else {
+		responseBody, err =
+			authenticatedGet("https://api.github.com/events?page="+strconv.Itoa(page), token)
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
 	var events GithubEvents
-	json.Unmarshal(body, &events)
+	json.Unmarshal(responseBody, &events)
 
 	return events, nil
 }
 
 // authenticatedGet performs an authenticated call to the GitHub's API using the user-supplied token.
 // A custom call to `unauthenticatedGet` is made under the hood.
-func authenticatedGet(pages int, page int, token string) (GithubEvents, error) {
+func authenticatedGet(url string, token string) ([]byte, error) {
 	ts := oauth2.StaticTokenSource(
 		&oauth2.Token{AccessToken: token},
 	)
 	tc := oauth2.NewClient(oauth2.NoContext, ts)
 
-	return unauthenticatedGet(pages, page, tc)
+	return unauthenticatedGet(url, tc)
 }
 
 // GetHubData returns a success/failure boolean (respectively `true`/`false`) along with the marshalled API data.
 // See the `GithubEvents` struct.
 func GetHubData(pages int, page int, token string) (GithubEvents, error) {
-	if token != "" {
-		return authenticatedGet(pages, page, token)
+	return getGithubEvents(pages, page, token)
+}
+
+// GetRateLimits fetches rate limits from GitHub API server for the current client. It returns a `rateLimitSpecs`
+// data structures containg reset timestamp, used requests, max requests and a default polling interval
+func GetRateLimits(token string) rateLimitSpecs {
+	// TODO: Check if rateLimitSpecs is already initialized
+
+	var response []byte
+	var err error
+
+	const rateLimitURL = "https://api.github.com/rate_limit"
```

Comment:
Here's the problem:

In order to set the refresh interval we need to get the rate limits from GitHub (see block `main.go:160-168`). Before, it was done in a `buildGraph()` call in line 158.
I wrote a separate public function so that we don't waste requests - the one to get the limits and the one to build the graph (requests to get the rate limits through the `rate_limit` endpoint don't count).

About the `rate` object: I didn't check the docs. Should've paid more attention in class.